bluetooth: controller: Use ticker must_expire to ensure ADV timing rand

mtpr-ot opened the pull request:
Enabled use of ticker must_expire feature for ensuring ADV timing
randomization, even when ADV doesn't get air-time. This reduces ADV
collisions. Active for non-nRF platforms only for now.

Signed-off-by: Morten Priess <mtpr@oticon.com>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -842,7 +842,14 @@ u8_t ll_adv_enable(u8_t enable)
+#if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE) && \
+	!defined(CONFIG_SOC_COMPATIBLE_NRF)
+				   /* Force expiry to ensure timing update */
+				   TICKER_LAZY_MUST_EXPIRE,
```
Just a reminder for me, is the TICKER_LAZY_MUST_EXPIRE supplied to the ticker_cb dynamic and used only when the event does not get the time reservation to use the radio?

mtpr-ot replied:
> Just a reminder for me, is the TICKER_LAZY_MUST_EXPIRE supplied to the ticker_cb dynamic and used only when the event does not get the time reservation to use the radio?

Yes, that's exactly how it works

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -842,7 +842,14 @@ u8_t ll_adv_enable(u8_t enable)
-				   TICKER_NULL_REMAINDER, TICKER_NULL_LAZY,
+				   TICKER_NULL_REMAINDER,
+#if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE) && \
+	!defined(CONFIG_SOC_COMPATIBLE_NRF)
```
I don't think this should depend on SOC_COMPATIBLE_NRF. What would break if that guard was removed?

mtpr-ot replied:
> I don't think this should depend on SOC_COMPATIBLE_NRF. What would break if that guard was removed?

I guess it would be OK to not have the !SOC_COMPATIBLE_NRF as dependency - and it makes good sense, as long as other platforms can accept the additional ticker updates when advertiser does not get air-time. @cvinayak, what is your opinion?

cvinayak replied:
its ok to not have !SOC_COMPATIBLE_NRF. Instead use CONFIG_BT_CTLR_LOW_LAT which covers nRF51, and will avoid this code block for nRF51

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -996,23 +1003,25 @@ static void ticker_cb(u32_t ticks_at_expire, u32_t remainder, u16_t lazy,
-	p.lazy = lazy;
-	p.param = lll;
-	mfy.param = &p;
+	if (lazy != TICKER_LAZY_MUST_EXPIRE) {
```
```if (IS_ENABLED(CONFIG_TICKER_COMPATIBILITY_MODE) ||  (lazy != ....)) {```

mtpr-ot committed:
bluetooth: controller: Use ticker must_expire to ensure ADV timing rand

Enabled use of ticker must_expire feature for ensuring ADV timing
randomization, even when ADV doesn't get air-time. This reduces ADV
collisions. Not active for nRF51 platform for now.

Signed-off-by: Morten Priess <mtpr@oticon.com>

cvinayak reviewed (approved):
LGTM
